python/cuml/dask/linear_model/linear_regression.py

```diff
@@ -88,217 +81,105 @@ def _func_create_model(sessionId, **kwargs):
+        Usually, you will not need to force co-locality if you pass the
+        X and y as follows;
 
-        key = uuid1()
-        partsToSizes = [(worker_info[wf[0]]["r"], self.client.submit(
-            LinearRegression._func_get_size,
-            wf[1],
-            workers=[wf[0]],
-            key="%s-%s" % (key, idx)).result())
-            for idx, wf in enumerate(X_futures)]
+        fit(X["all_the_columns_but_labels"], X["labels"])
 
-        n_cols = X.shape[1]
-        n_rows = reduce(lambda a, b: a+b, map(lambda x: x[1], partsToSizes))
+        You might want to force co-locality if you pass the X and y as
+        follows;
 
-        key = uuid1()
-        self.linear_models = [(wf[0], self.client.submit(
-            LinearRegression._func_create_model,
-            comms.sessionId,
-            **self.kwargs,
-            workers=[wf[0]],
-            key="%s-%s" % (key, idx)))
-            for idx, wf in enumerate(worker_to_parts.items())]
+        fit(X, y)
 
-        key = uuid1()
-        linear_fit = dict([(worker_info[wf[0]]["r"], self.client.submit(
-            LinearRegression._func_fit,
-            wf[1],
-            worker_to_parts[wf[0]],
-            worker_to_parts_y[wf[0]],
-            n_rows, n_cols,
-            partsToSizes,
-            worker_info[wf[0]]["r"],
-            key="%s-%s" % (key, idx),
-            workers=[wf[0]]))
-            for idx, wf in enumerate(self.linear_models)])
+        because dask might have distributed the partitions of X and y
+        into different workers.
 
-        wait(list(linear_fit.values()))
-        raise_exception_from_futures(list(linear_fit.values()))
+        Parameters
+        ----------
+        X : dask cuDF dataframe (n_rows, n_features)
+            Features for regression
+        y : dask cuDF (n_rows, 1)
+            Labels (outcome values)
+        force_colocality: boolean (True: re-distributes the partitions
+                          of X and y to force the co-locality of
+                          the partitions)
+        """
 
-        comms.destroy()
+        # todo: add check for colocality in case force_colocality=False
 
-        self.local_model = self.linear_models[0][1].result()
-        self.coef_ = self.local_model.coef_
-        self.intercept_ = self.local_model.intercept_
+        X = self.client.persist(X)
+        y = self.client.persist(y)
 
-    def _fit_with_colocality(self, X, y):
-        input_futures = self.client.sync(extract_colocated_ddf_partitions,
-                                         X, y, self.client)
-        workers = list(input_futures.keys())
+        data = MGData.colocated(data=(X, y), client=self.client)
+        self.datatype = data.datatype
 
         comms = CommsContext(comms_p2p=False)
-        comms.init(workers=workers)
+        comms.init(workers=data.workers)
 
-        worker_info = comms.worker_info(comms.worker_addresses)
+        data.calculate_parts_to_sizes(comms)
```

Comment:
This line seems a little weird to me... does it just compute the sizes and store them in `data.something` as a side effect? Seems hard to parse the dataflow here.... would be nice to explicitly return the info we need from this calculation or do it all automatically an internally to MGData when needed.